### DISCUSSION TITLE
Cannot get quotes from YahooFinance

### DISCUSSION THREAD

**Discussion by SitramSoft** [ghostfolio/ghostfolio#3848] [Q&A]
**Bug Description**

After a certain period of time, I cannot get quotes from YahooFinance when I access the app from the browser when running Ghostfolio in docker.

**To Reproduce**

I run Ghostfolio in Docker so I can access it remotely trough a subdomain I own. This means the container is running 24/7(minus eventual downtimes).

After a certain period of time, I cannot see quotes for the symbols I have defined and docker logs show a lot of messages similar with the one below:

**Expected behavior**

Quotes for symbols are always loaded correctly.

**Screenshots**

N/A

**Logs**

```
[Nest] 260  - 09/18/2024, 12:29:43 PM     LOG [PortfolioSnapshotProcessor (PORTFOLIO)] Portfolio snapshot calculation of user 'xxx' has been started
[Nest] 260  - 09/18/2024, 12:29:43 PM    WARN [YahooFinanceService] Fallback to yahooFinance.quoteSummary()
[Nest] 260  - 09/18/2024, 12:29:43 PM   ERROR [YahooFinanceService] FetchError: request to https://guce.yahoo.com/consent?brandType=nonEu&gcrumb=XMWcb28&done=https%3A%2F%2Ffinance.yahoo.com%2Fquote%2FAAPL failed, reason: Hostname/IP does not match certificate's altnames: Host: guce.yahoo.com. is not in the cert's altnames: DNS:app.groenopruiming.com, DNS:groenopruiming.com
[Nest] 260  - 09/18/2024, 12:29:43 PM   ERROR [YahooFinanceService] Could not get quote summary for IBKR
[Nest] 260  - 09/18/2024, 12:29:43 PM   ERROR [YahooFinanceService] Could not get quote summary for NFLX
[Nest] 260  - 09/18/2024, 12:29:43 PM   ERROR [YahooFinanceService] Could not get quote summary for NVDA
[Nest] 260  - 09/18/2024, 12:29:43 PM     LOG [PortfolioSnapshotProcessor (PORTFOLIO)] Portfolio snapshot calculation of user 'xxx' has been completed in 0.264 seconds
[Nest] 260  - 09/18/2024, 12:29:53 PM    WARN [YahooFinanceService] Fallback to yahooFinance.quoteSummary()
[Nest] 260  - 09/18/2024, 12:29:53 PM   ERROR [YahooFinanceService] FetchError: request to https://guce.yahoo.com/consent?brandType=nonEu&gcrumb=XMWcb28&done=https%3A%2F%2Ffinance.yahoo.com%2Fquote%2FAAPL failed, reason: Hostname/IP does not match certificate's altnames: Host: guce.yahoo.com. is not in the cert's altnames: DNS:app.groenopruiming.com, DNS:groenopruiming.com
[Nest] 260  - 09/18/2024, 12:29:53 PM   ERROR [YahooFinanceService] Could not get quote summary for ^SPX
[Nest] 260  - 09/18/2024, 12:29:53 PM   ERROR [ExceptionsHandler] Cannot read properties of undefined (reading 'currency')
TypeError: Cannot read properties of undefined (reading 'currency')
    at BenchmarkService.getMarketDataForUser (/ghostfolio/apps/api/main.js:1:328723)
```

After restarting the docker container I get the following messages in log and the quotes for all my symbols are correctly shown.

```
[Nest] 260  - 09/18/2024, 12:42:39 PM     LOG [NestApplication] Nest application successfully started +130ms
[Nest] 260  - 09/18/2024, 12:42:39 PM     LOG    ________               __  ____      ___
[Nest] 260  - 09/18/2024, 12:42:39 PM     LOG   / ____/ /_  ____  _____/ /_/ __/___  / (_)___
[Nest] 260  - 09/18/2024, 12:42:39 PM     LOG  / / __/ __ \/ __ \/ ___/ __/ /_/ __ \/ / / __ \
[Nest] 260  - 09/18/2024, 12:42:39 PM     LOG / /_/ / / / / /_/ (__  ) /_/ __/ /_/ / / / /_/ /
[Nest] 260  - 09/18/2024, 12:42:39 PM     LOG \____/_/ /_/\____/____/\__/_/  \____/_/_/\____/ 2.108.0
[Nest] 260  - 09/18/2024, 12:42:39 PM     LOG 
[Nest] 260  - 09/18/2024, 12:42:39 PM     LOG Listening at http://0.0.0.0:3333
[Nest] 260  - 09/18/2024, 12:42:39 PM     LOG 
[Nest] 260  - 09/18/2024, 12:44:30 PM   ERROR [ExchangeRateDataService] No exchange rate has been found for EURUSD
[Nest] 260  - 09/18/2024, 12:44:30 PM   ERROR [ExchangeRateDataService] No exchange rate has been found for EURUSD
[Nest] 260  - 09/18/2024, 12:44:30 PM     LOG [PortfolioSnapshotProcessor (PORTFOLIO)] Portfolio snapshot calculation of user 'xxx' has been started
Please consider completing the survey at https://bit.ly/yahoo-finance-api-feedback if you haven't already; for more info see https://github.com/gadicc/node-yahoo-finance2/issues/764#issuecomment-2056623851.
Fetching crumb and cookies from https://finance.yahoo.com/quote/AAPL...
fetch https://guce.yahoo.com/consent?brandType=nonEu&gcrumb=Cwz-hUc&done=https%3A%2F%2Ffinance.yahoo.com%2Fquote%2FAAPL
fetch https://consent.yahoo.com/v2/collectConsent?sessionId=3_cc-session_3532e590-7d5d-46eb-9fd0-621d5a25c15a
fetch https://consent.yahoo.com/v2/collectConsent?sessionId=3_cc-session_3532e590-7d5d-46eb-9fd0-621d5a25c15a
fetch https://guce.yahoo.com/copyConsent?sessionId=3_cc-session_3532e590-7d5d-46eb-9fd0-621d5a25c15a&lang=ro-RO
Fetching crumb and cookies from https://finance.yahoo.com/quote/AAPL?guccounter=1...
We expected a redirect to guce.yahoo.com, but got https://finance.yahoo.com/quote/AAPL/
We'll try to continue anyway - you can safely ignore this if the request succeeds
Success. Cookie expires on Infinity
fetch https://query1.finance.yahoo.com/v1/test/getcrumb
New crumb: A1eKi3Ko9wO
[Nest] 260  - 09/18/2024, 12:44:31 PM   ERROR [YahooFinanceService] Error: User is unable to access this feature - https://bit.ly/yahoo-finance-api-feedback
[Nest] 260  - 09/18/2024, 12:44:31 PM    WARN [YahooFinanceService] Fallback to yahooFinance.quoteSummary()
[Nest] 260  - 09/18/2024, 12:44:31 PM   ERROR [YahooFinanceService] Could not get quote summary for IBKR
[Nest] 260  - 09/18/2024, 12:44:32 PM     LOG [PortfolioSnapshotProcessor (PORTFOLIO)] Portfolio snapshot calculation of user 'xxx' has been completed in 2.070 seconds
[Nest] 260  - 09/18/2024, 12:44:32 PM   ERROR [ExchangeRateDataService] No exchange rate has been found for EURUSD
[Nest] 260  - 09/18/2024, 12:44:32 PM   ERROR [ExchangeRateDataService] No exchange rate has been found for EURUSD
[Nest] 260  - 09/18/2024, 12:44:32 PM   ERROR [ExchangeRateDataService] No exchange rate has been found for EURUSD
[Nest] 260  - 09/18/2024, 12:44:32 PM   ERROR [ExchangeRateDataService] No exchange rate has been found for EURUSD
[Nest] 260  - 09/18/2024, 12:44:32 PM   ERROR [ExchangeRateDataService] No exchange rate has been found for EURUSD
[Nest] 260  - 09/18/2024, 12:44:32 PM   ERROR [ExchangeRateDataService] No exchange rate has been found for EURUSD
[Nest] 260  - 09/18/2024, 12:44:32 PM   ERROR [BenchmarkService] No historical market data has been found for ^SPX (YAHOO) at 2023-12-31
[Nest] 260  - 09/18/2024, 12:45:16 PM   ERROR [ExchangeRateDataService] No exchange rate has been found for EURUSD
[Nest] 260  - 09/18/2024, 12:45:16 PM   ERROR [ExchangeRateDataService] No exchange rate has been found for EURUSD
```

**Environment**

My docker-compose.yaml:

```docker
ghostfolio:
    image: ghostfolio/ghostfolio:latest
    container_name: Ghostfolio
    env_file:
      - .env
    environment:
      DATABASE_URL: ${DATABASE_URL}
      NODE_ENV: production
      REDIS_HOST: ${REDIS_HOST}
      REDIS_PORT: ${REDIS_PORT}
    ports:
      - 3333:3333
    restart: unless-stopped
```

DATABASE_URL has the following value. Each of the remaining variables are my user, host and password for the Postgres docker container.

- Ghostfolio Version 2.108.0
- Self-hosted
- Experimental Features enabled
- Browser: Chrome Version 128.0.6613.119 (Official Build) (64-bit)
- OS: Arch Linux

**Additional context**

N/A

**Comment by SitramSoft** [comment]
It's been over 16 hours since my last container restart and I still can get quotes from YahooFinance. I'll continue to monitor and update this bug report.

**Comment by SDuesterhaupt** [comment]
@SitramSoft 

Are you running your container in Docker rootless mode?

When the setup was moved to rootless mode, the necessary rules in iptables for communication from the inside to the outside probably were not set automatically.

Solution (tested in RockyLinux9):

In the file ```~/.config/systemd/user/docker.service``` convert the line

```ExecStart=/home/containeradmin/bin/dockerd-rootless.sh --iptables=false```

to

```ExecStart=/home/containeradmin/bin/dockerd-rootless.sh```.

Finally, reload the systemd daemon:

```systemctl --user daemon-reload```

And restart Docker:

```systemctl --user restart docker```

**Comment by SitramSoft** [comment]
I am not running docker in rootless mode.

The server where this container runs, doesn't have any firewall rules set, because it is in the DMZ zone of my home lab. All the other containers have no issue in accessing the internet. Even this one accesses it correctly for a while, then something happens and it doesn't work.

**Comment by SDuesterhaupt** [comment]
Okay, then you should take a look onto you DNS resolving. Probably you run a DNS filter (pihole etc.) and the Yahoo domain is somehow blacklisted after an update.

Just log into the running container an analyze the way of resolving.

Btw. ... It's a DMZ host... why do you bind your stack on 3333:3333? So if you run Docker in root mode then you drill a hole into your firewall where the world can potentially access.

Better: '127.0.0.1:3333:3333' and/or '[::1]:3333:3333' and use a reverse proxy in front of the stack.

**Comment by SitramSoft** [comment]
@SDuesterhaupt 

I doubt there is an issue with my DNS because I use the same rules for my entire house and the container runs on the same host as 34 other containers, all of them using the same DNS server to resolve any queries they need.

Here is an snippet from the console of this container with the result of the dig command on  finance.yahoo.com. Seems the host is correctly resolved.
![image](https://github.com/user-attachments/assets/72ee8245-bd87-4d81-a871-601240690aec)

Maybe something messes up after the container runs for a day or so. I can do another ping tomorrow after the container ran for about a day. Before doing this test, I installed new updates on the server, including an update for docker. This lead to restarting all my containers, so Ghostfolio works correctly.

**Comment by SitramSoft** [comment]
I tried manually downloading the page from the link that causes the error in my first post and I don't get any error from the console of the container.

![image](https://github.com/user-attachments/assets/f1697ed6-cb79-42fe-86d8-fa7f0aea4fb6)

**Comment by SDuesterhaupt** [comment]
> Maybe something messes up after the container runs for a day or so. I can do another ping tomorrow after the container ran for about a day. Before doing this test, I installed new updates on the server, including an update for docker. This lead to restarting all my containers, so Ghostfolio works correctly.

Look at the Docker log next time you get a bug, as developers are still working on the productive implementation of IPv6. The proxy is still good for surprises, and there have been a lot of updates lately, at least regarding the CE.

Then it may also be that the interaction of iptables/firewalld, and fail2ban occasionally overwrite Docker rules (bridges between container and host using the Docker's IP subnet area) into the iptables. This can be fixed by restarting firewalld (1st), fail2ban (2nd) and docker (3rd).

I would say that the restart of the Docker-Service has finally fixed the bug, and Ghostfolio itself works at this point.

**Reply by SitramSoft** [reply]
I appreciate the willingness to help, but I don't understand if the suggestions you offer are generic or you actually took the time to look over the information I posted. 

You suggested to looked at the docker log next time. This is what I did in the first post. I even included a snippet from the log. Looking at the error that appears in the log after the container runs for a day, I can see that there is an issue with the certificate altname. Host: `guce.yahoo.com` is not in the cert's altnames: `DNS:app.groenopruiming.com, DNS:groenopruiming.com`

`ERROR [YahooFinanceService] FetchError: request to https://guce.yahoo.com/consent?brandType=nonEu&gcrumb=XMWcb28&done=https%3A%2F%2Ffinance.yahoo.com%2Fquote%2FAAPL failed, reason: Hostname/IP does not match certificate's altnames: Host: guce.yahoo.com. is not in the cert's altnames: DNS:app.groenopruiming.com, DNS:groenopruiming.com`

How did you draw the conclusion that the issue is with DNS resolving, since clearly the container can resolve the DNS query and I have done additional tests inside the container by installing additional tools to check your original assumption that there is an issue with DNS resolving.

> I would say that the restart of the Docker-Service has finally fixed the bug, and Ghostfolio itself works at this point.

Restarting the docker container, not the docker-service `temporary` fixes the issue. A day later I experience the same symptom. I don't think the app creator intended to have the container restarted every day. This is a bug and I did my contribution to open source to report it.

**Reply by SDuesterhaupt** [reply]
Okay, I seem to waste your time... short form... I always read a post in detail and I try to figure out the message.

1. Docker log vs. stack/container log
> You suggested to looked at the docker log next time. This is what I did in the first post.

No, you didn't do. The log of the ```docker.service``` (runtime environment/dockerd, e.g. ```journalctl -xefu docker.service```) is not the log inside the stack (e.g. ```docker log -f ghostfolio```). Different worlds.

2. DNS vs. certificate
> How did you draw the conclusion that the issue is with DNS resolving...

Receiving this warning ```Hostname/IP does not match certificate's altnames``` has only one reason. The answering host's domain name is not the same as in the certificate. The trusted chain check fails. TLS expects ```guce.yahoo.com``` and the domain is something for instance like ```https://172.x.x.x/consent?...``` or ```https://mylocal.net/consent?...``` due to issues during proxying (userland proxy IPv4/IPv6) or local DNS resolving. Keyword: Man-in-the-Middle.

3. Again docker.service vs. docker container > Restarting
> Restarting the docker container, not the docker-service temporary fixes the issue.

In a former post you wrote
> I installed new updates on the server, including an update for docker. This lead to restarting all my containers.

A system update of dockerd (docker damon) leads often to an automatic restart of the dockerd (e.g. ```systemctl restart docker.service```). With respect of the quote I concluded you updated the docker-framework (and not the ghostfolio-/db-/whatever-image inside the ghostfolio stack). If so then restarting dockerd leads to a reset of iptable rules regarding dockerd (forwarding between inside world and outside world using defined bridges).

Everything else makes no or less sense. Updating something inside the ghostfolio stack during its runtime is not persistent. And I guess you didn't build your own ghostfolio-/db-/whatever-image, did you?

4. 24 h
> A day later I experience the same symptom.

So it's systematic and repeatable. It's sounds rather like a (lease) or caching issue within your DHCP/DNS/Pihole/Unbound solution.

--

The issue is not the ghostfolio stack itself as long as you do the setup as suggested by the developers. Otherwise developers or users like you and me would simply jump into that discussion within the last days. I jumped in due to the error messages of the container. In my case within a rootlesskit setup the building of the iptables chain failed (limited access to iptables because of user with restricted permissions). My solution I just wanted to share.

Have a nice time... That's my (last) 2 cent.

**Reply by SitramSoft** [reply]
First of all I'd like to apologize for my tone in the previous reply. I know you were just trying to help and I shouldn't have replied so harsh.

I did another test today on the container and the market data for the symbols that I have stored in my account are updating. I tried to add another symbol(GOOGL) and the search field successfully retrieved the symbol from Yahoo Finance. I also don't see anymore the same errors related to certificates in the docker container log, that appeared when I first open the issue on GitHub.

I'll keep monitoring in the next couple of days to see if this issue still appears or not.

**Comment by SDuesterhaupt** [comment]
> I doubt there is an issue with my DNS because I use the same rules for my entire house and the container runs on the same host as 34 other containers, all of them using the same DNS server to resolve any queries they need.

And a clarification regarding DNS in Docker...

I was referring to the Docker internal DNS server (127.0.0.11:53, as you can see in your screenshot below 'QUERY TIME').

Container/DNS request for example321.com > Docker/127.0.0.11:53 UDP > if necessary forwarding to your internal DNS > if necessary forwarding to external DNS

**Reply by SitramSoft** [reply]
I have a dedicated DHCP server which is configured to set as a LAN wide DNS the address of the VM server that runs Pihole and Unbound. I am not an expert in how Docker resolves a DNS query, but I expect that unless a special network is configured, there is a separate private network created for each docker instance with the host and all DNS queries to Docker/127.0.0.11:53 UDP are in fact forwarded by the host OS to the DSN server it has received when performing the DHCP authentication at startup.
